### PR TITLE
docs - add pxf config info for s3 server-side encryption

### PR DESCRIPTION
**Pull request by lisakowen**
PXF s3 connector supports s3 server-side encryption.  describe how to configure SSE.  (recommend setting bucket policies, also describe how to set via pxf server s3-site.xml configuration file.)

question:  if a default bucket policy is set for bucket1, SSE is also configured in s3-site.xml, and a writable external table is used to write to bucket1 - which SSE config is used?

doc review site link:
- http://docs-lisa-pxfs3sse.cfapps.io/6-0/pxf/objstore_cfg.html#s3_cfg

### CODE REVIEW

**Comment by kongyew** [comment]
@denalex Can you please answer Lisa 's question about "if a default bucket policy is set for bucket1, SSE is also configured in s3-site.xml, and a writable external table is used to write to bucket1 - which SSE config is used?"

**Comment by denalex** [comment]
I'd think that the configured SSE for the client (s3-site.xml) will take precedence over the default bucket SSE, however to be 100% sure we would need to test it out.

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 1]
```diff
@@ -113,6 +113,109 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+  <name>fs.s3a.server-side-encryption-algorithm</name>
+  <value>AES256</value>
+</property>
+```
+
+To enable SSE-S3 for a specific S3 bucket, use the property name variant that includes the bucket name. For example:
+
+``` xml
+<property>
+  <name>fs.s3a.bucket.YOUR_BUCKET1_NAME.server-side-encryption-algorithm</name>
```
[the remark this answers is not in the document]
Comment (lisakowen) [reply gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: yes, i added a **1** to this bucket name example and a **2** to the KMS bucket-specific version examples to distinguish different buckets.  i can remove the numbers if confusing.

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 2]
```diff
@@ -113,6 +113,109 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+  <value>SSE-KMS</value>
+</property>
+<property>
+  <name>fs.s3a.server-side-encryption.key</name>
+  <value>YOUR_AWS_SSE_KMS_KEY_ARN</value>
+</property>
+```
+
+Substitute `YOUR_AWS_SSE_KMS_KEY_ARN` with your key resource name. If you do not specify an encryption key, the default key defined in the Amazon KMS is used. Example KMS key: `arn:aws:kms:us-west-2:123456789012:key/1a23b456-7890-12cc-d345-6ef7890g12f3`.
+
```
[the remark this answers is not in the document]
Comment (lisakowen) [reply gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: will add this.  thanks!

**Comment by lisakowen** [comment]
review site has been updated.

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 3]
```diff
@@ -111,8 +111,115 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
 
+
+#### <a id="s3-sse"></a>Configuring S3 Server-Side Encryption
+
+PXF supports Amazon Web Service S3 Server-Side Encryption (SSE) for S3 files that you access with readable and writable Greenplum Database external tables that specify the `pxf` protocol and an `s3:*` profile. AWS S3 server-side encryption protects your data at rest; it encrypts your object data as it writes to disk, and transparently decrypts the data for you when you access it.
+
+PXF supports the following AWS SSE encryption key management schemes:
+
+- SSE with S3-Managed Keys (SSE-S3) - Amazon manages the data and master encryption keys.
+- SSE with Key Management Service Managed Keys (SSE-KMS) - Amazon manages the data key, you manage the encryption key in AWS KMS.
```
Comment (dyozie) [inline gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: key, you -> key, and you

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 4]
```diff
@@ -111,8 +111,115 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+- SSE with Customer-Provided Keys (SSE-C) - You set and manage the encryption key.
+
+Your S3 access key and secret key govern your access to all S3 bucket objects, whether the data is encrypted or not.
+
+S3 transparently decrypts data during a read operation of an encrypted file that you access via a readable external table created specifying the `pxf` protocol and an `s3:*` profile. No additional configuration is required.
```
Comment (dyozie) [inline gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: table created specifying -> table that is created by specifying

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 5]
```diff
@@ -111,8 +111,115 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+
+To encrypt data that you write to S3 via a writable external table created specifying the `pxf` protocol and an `s3:*` profile, you have two options, discussed below:
```
Comment (dyozie) [inline gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: I think we can leave out the full definition in this second mention.  Maybe just edit to:  "To encrypt data that you write to S3 via this type of external table, you have two options:"

Comment (dyozie) [reply gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: Either that or say "a supported external table"

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 6]
```diff
@@ -111,8 +111,115 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+- Configure SSE encryption options in your PXF S3 server `s3-site.xml` configuration file.
+
+##### <a id="s3-sse-bucket"></a>Configuring SSE via an S3 Bucket Policy (Recommended)
+
+You can create S3 <i>Bucket Policy</i>(s) that identify the objects that you want to encrypt, the encryption key management scheme, and the write actions permitted on those objects. Refer to [Protecting Data Using Server-Side Encryption](http://docs.aws.amazon.com/AmazonS3/latest/dev/serv-side-encryption.html) in the AWS S3 documentation for more information about the SSE encryption key management schemes. [How Do I Enable Default Encryption for an S3 Bucket?](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/default-bucket-encryption.html) describes how to set default encryption bucket policies.
+
+
+##### <a id="s3-sse-pxfs3cfg"></a>Specifying SSE Options in a PXF S3 Server Configuration
+
+You must include certain properties in `s3-site.xml` to configure server-side encryption in a PXF S3 server configuration. The properties and values that you add to the file are SSE encryption key management scheme-specific.
```
Comment (dyozie) [inline gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: are SSE encryption key management scheme-specific. -> dependent on the SSE encryption key management scheme.

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 7]
```diff
@@ -111,8 +111,115 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+  <name>fs.s3a.bucket.YOUR_BUCKET_NAME.server-side-encryption-algorithm</name>
+  <value>AES256</value>
+</property>
+```
+
+Replace `YOUR_BUCKET_NAME` with the name of the S3 bucket.
+
+**SSE-KMS**
+
+To enable SSE-KMS on any file that you write to any S3 bucket, you must set both the encryption algorithm and encryption key ID. Set these properties in the `s3-site.xml` file:
```
Comment (dyozie) [inline gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: you must set -> set

Comment (dyozie) [reply gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: Set these properties -> To set these properties

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 8]
```diff
@@ -111,8 +111,115 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+  <name>fs.s3a.bucket.YOUR_BUCKET_NAME.server-side-encryption.key</name>
+  <value>YOUR_AWS_SSE_KMS_KEY_ARN</value>
+</property>
+```
+
+Replace `YOUR_BUCKET_NAME` with the name of the S3 bucket.
+
+**SSE-C**
+
+To enable SSE-C on any file that you write to any S3 bucket, you must set both the encryption algorithm and the encryption key (base-64 encoded). All clients must share the same key.
```
Comment (dyozie) [inline gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: you must set -> set

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (kongyew) [reply gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: It is reasonable to differentiate both buckets .

File: gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb [hunk 9]
```diff
@@ -113,6 +113,109 @@ The template configuration file for S3 is `$PXF_CONF/templates/s3-site.xml`. Whe
...
+
+#### <a id="s3-sse"></a>Configuring S3 Server-Side Encryption
+
+PXF supports Amazon Web Service S3 Server-Side Encryption (SSE) for S3 files that you access with readable and writable Greenplum Database external tables that specify the `pxf` protocol and an `s3:*` profile. AWS S3 server-side encryption protects your data at rest; it encrypts your object data as it writes to disk, and transparently decrypts the data for you when you access it.
+
+PXF supports the following AWS SSE encryption key management schemes:
+
+- SSE with S3-Managed Keys (SSE-S3) - Amazon manages the data and master encryption keys.
+- SSE with Key Management Service Managed Keys (SSE-KMS) - Amazon manages the data key, you manage the encryption key in AWS KMS.
+- SSE with Customer-Provided Keys (SSE-C) - You set and manage the encryption key.
```
[the remark this answers is not in the document]
Comment (kongyew) [reply gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb]: Yes,  it is reasonable to claim all  SSE features are supported by AWS-S3

We tested those modes, so we can understand any issues with usabilities and frictions.  If there are any issues with this last mode, AWS-S3 or Hadoop Compatible file system for S3 is most likely the root cause.

**Comment by kongyew** [comment]
Reviewed

**Pull request merged**